artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerConsumerImpl.java

```diff
@@ -1513,7 +1524,137 @@ public String getConnectionRemoteAddress() {
+   }
+
+   @Override
+   public long getLastAcknowledgedTime() {
+      return metrics.getLastAcknowledgedTime();
+   }
+
+   @Override
+   public long getMessagesAcknowledged() {
+      return metrics.getMessagesAcknowledged();
+   }
+
+   @Override
+   public int getMessagesDeliveredSize() {
+      return metrics.getMessagesDeliveredSize();
+   }
+
+   @Override
+   public int getMessagesDelivered() {
+      return metrics.getMessagesDelivered();
+   }
+
+   @Override
+   public int getMessagesAcknowledgedAwaitingCommit() {
+      return metrics.getMessagesAcknowledgedAwaitingCommit();
+   }
+
    public SessionCallback getCallback() {
       return callback;
    }
+
+   static class ServerConsumerMetrics extends TransactionOperationAbstract {
+
+      /**
+       * Since messages can be delivered (incremented) and acknowledged (decremented) at the same time we have to protect
+       * the encode size and make it atomic. The other fields are ok since they are only accessed from a single thread.
+       */
+      private static final AtomicIntegerFieldUpdater<ServerConsumerMetrics> messagesInTransitSizeUpdater = AtomicIntegerFieldUpdater.newUpdater(ServerConsumerMetrics.class, "messagesInTransitSize");
+
+      private volatile int messagesInTransitSize = 0;
+
+      private static final    AtomicIntegerFieldUpdater<ServerConsumerMetrics> messagesAcknowledgedAwaitingCommitUpdater = AtomicIntegerFieldUpdater.newUpdater(ServerConsumerMetrics.class, "messagesAcknowledgedAwaitingCommit");
+
+      private volatile int messagesAcknowledgedAwaitingCommit = 0;
+
+      private int messagesDeliveredSize = 0;
+
+      private long lastDeliveredTime = 0;
+
+      private long lastAcknowledgedTime = 0;
+
+      private int messagesDelivered = 0;
+
+      private int messagesAcknowledged = 0;
+
+
+      public int getMessagesInTransitSize() {
+         return messagesInTransitSizeUpdater.get(this);
+      }
+
+      public int getMessagesDeliveredSize() {
+         return messagesDeliveredSize;
+      }
+
+      public long getLastDeliveredTime() {
+         return lastDeliveredTime;
+      }
+
+      public long getLastAcknowledgedTime() {
+         return lastAcknowledgedTime;
+      }
+
+      public int getMessagesDelivered() {
+         return messagesDelivered;
+      }
+
+      public int getMessagesAcknowledged() {
+         return messagesAcknowledged;
+      }
+
+      public int getMessagesAcknowledgedAwaitingCommit() {
+         return messagesAcknowledgedAwaitingCommitUpdater.get(this);
+      }
+
+      public void addMessage(int encodeSize) {
+         messagesInTransitSizeUpdater.addAndGet(this, encodeSize);
+         messagesDeliveredSize += encodeSize;
+         messagesDelivered++;
+         lastDeliveredTime = System.currentTimeMillis();
+      }
+
+      public void addAcknowledge(int encodeSize, Transaction tx) {
+         messagesInTransitSizeUpdater.addAndGet(this, -encodeSize);
+         messagesAcknowledged++;
```

Comment:
I think this one is unsafe also. Looks like rollback synchronizes 'lock' but not the consumer itself as some other calls of this method from other threads do. The openwire bit using it doesnt synchronize on either.

Given the large number of places this method is used from, some on the queue thread, others not, and the mishmash of different synchronization by those calls...I think it would be simpler/safer to just use an updater for this regardless whether it is currently safe, it is certainly brittle and its likely someone change other bits in a way that breaks this area later if it isnt already (which it looks like it is).

Reply:
do you mean the messagesAcknowledged field?

Reply:
Yes. The method seems like it can still be multi-threaded, so doing ++ isnt safe.